pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
-// public key is a public key of the signing group. It publishes the public key
-// to the keep. It uses keep address as unique signing group identifier.
+// GenerateSignerForKeep generates a new threshold signer with ECDSA key pair
+// and submits the public key to the on-chain keep.
+//
+// The attempt for generating signer is retried on failure until the provided
+// context is done.
 func (n *Node) GenerateSignerForKeep(
 	ctx context.Context,
 	operatorPublicKey *operator.PublicKey,
 	keepAddress common.Address,
-	keepMembersIDs []tss.MemberID,
+	members []common.Address,
 ) (*tss.ThresholdSigner, error) {
 	memberID := tss.MemberIDFromPublicKey(operatorPublicKey)
 
-	signer, err := tss.GenerateThresholdSigner(
-		ctx,
-		keepAddress.Hex(),
-		memberID,
-		keepMembersIDs,
-		uint(len(keepMembersIDs)-1),
-		n.networkProvider,
-		n.tssParamsPool.get(),
-	)
-	if err != nil {
-		return nil, fmt.Errorf("failed to generate threshold signer: [%v]", err)
+	logger.Infof("Getting new key-generation pre-parameters from the pool")
+	preParamsBox := params.NewBox(n.tssParamsPool.get())
+	logger.Infof("Received new key-generation parameters from the pool")
+
+	attemptCounter := 0
+	for {
+		attemptCounter++
+
+		logger.Infof(
+			"signer generation for keep [%s]; attempt [%v]",
+			keepAddress.String(),
+			attemptCounter,
+		)
+
+		// If we are re-attempting the key generation, pre-parameters in the box
+		// could be destroyed because they were shared with other members.
+		// In this case, we need to re-generate them.
+		if preParamsBox.IsEmpty() {
+			logger.Infof("Getting new key-generation pre-parameters from the pool")
+			preParamsBox = params.NewBox(n.tssParamsPool.get())
+			logger.Infof("Received new key-generation parameters from the pool")
+		}
+
+		// Global timeout for generating a signer exceeded.
+		// We are giving up and leaving this function.
+		if ctx.Err() != nil {
+			return nil, fmt.Errorf("key generation timeout exceeded")
+		}
+
+		// Announce signer presence. Other members of the keep need to receive
+		// the public key of this members. This member, need to receive public
+		// keys of all other members. Up to this point, only addresses from
+		// signer selection protocol are known.
+		//
+		// If signer announcement fails, we retry from the beginning.
+		memberIDs, err := n.AnnounceSignerPresence(
+			ctx,
+			operatorPublicKey,
+			keepAddress,
+			members,
+		)
+		if err != nil {
+			logger.Errorf("failed to announce signer presence: [%v]", err)
+			continue
+		}
+
+		// Generate threshold signer by generating threshold key with all other
+		// keep members.
+		//
+		// If threshold key generation fails, we retry from the beginning.
+		signer, err := tss.GenerateThresholdSigner(
+			ctx,
+			keepAddress.Hex(),
+			memberID,
+			memberIDs,
+			uint(len(memberIDs)-1),
+			n.networkProvider,
+			preParamsBox,
+		)
+		if err != nil {
+			logger.Errorf("failed to generate threshold signer: [%v]", err)
```

Comment:
Since we're in a retry loop maybe we should log just a warning, not an error.

Reply:
Same explanation as for the error<>warning comment before.